native-heapdump-examples/README.md

```diff
@@ -0,0 +1,129 @@
+# Creating a Heap Dump from a Native Executable
+
+You can create a heap dump of a running executable to monitor its execution. Just like any other Java heap dump, it can be opened with the [VisualVM](https://www.graalvm.org/latest/tools/visualvm/) tool. To enable heap dump support, native executables must be built with the --enable-monitoring=heapdump option.
+
+Then you can request heap dumps in the same way you can request them when your application runs on the JVM (for example, right-click on the process, then select "Heap Dump"). There are different ways to create heap dumps:
+* Create heap dumps with VisualVM.
+* Dump the initial heap of a native executable using the `-XX:+DumpHeapAndExit` command-line option.
+* Create heap dumps sending a `SIGUSR1` signal at run time.
+* Create heap dumps programmatically using the [`org.graalvm.nativeimage.VMRuntime#dumpHeap`](https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/VMInspectionOptions.java) API.
+
+All approaches are described below.
+
+>Note: Creating heap dumps is not available on the Microsoft Windows platform.
+
+## Preparation
+
+1. Download and install the latest GraalVM JDK with Native Image and VisualVM support using the [GraalVM JDK Downloader](https://github.com/graalvm/graalvm-jdk-downloader):
+    ```bash
+    bash <(curl -sL https://get.graalvm.org/jdk) -c 'visualvm'
+    ```
+
+2. Download the demos repository or clone it as follows:
+    ```bash
+    git clone https://github.com/graalvm/graalvm-demos
+    ```
+    ```bash
+    cd graalvm-demos/native-heapdump-examples
+    ```
+
+## Dump the Initial Heap of a Native Executable
+
+Use the `-XX:+DumpHeapAndExit` command-line option to dump the initial heap of a native executable.
+This can be useful to identify which objects the `native-image` build process allocated to the executable's heap. 
+For a HelloWorld example, use the option as follows:
+
+```shell
+$GRAALVM_HOME/bin/native-image HelloWorld --enable-monitoring=heapdump
+```
+```shell
+./helloworld -XX:+DumpHeapAndExit
+```
+Heap dump created at '/path/to/helloworld.hprof'.
```

Comment:
The heap dump is created in the application's working directory, _/path/to/helloworld.hprof_.